### ISSUE TITLE
"Use of uninitialized value in substitution iterator" caused by mpt-status error

### ISSUE THREAD

**Issue by ebourg** [glensc/nagios-plugin-check_raid#95]
Hi,

On my system mpt-status returns the error message `ioctl: No such device`, it seems check_raid doesn't handle this case and displays this warning:

    Use of uninitialized value in substitution iterator at ./check_raid.pl line 430.

The uninitialized value is `$id`.

**Comment by glensc** [comment]
please see CONTRIBUTING.md in the repository how to report bugs. need debug data from your system to fix.

**Comment by ebourg** [comment]
Ok, I ran `./check_raid.pl -d` and got this:

    DEBUG EXEC: /proc/mdstat at ./check_raid.pl line 449.
    DEBUG EXEC: /sbin/MegaCli -PDList -aALL -NoLog at ./check_raid.pl line 449.
    DEBUG EXEC: /sbin/MegaCli -LdInfo -Lall -aALL -NoLog at ./check_raid.pl line 449.
    DEBUG EXEC: /sbin/mpt-status -p at ./check_raid.pl line 449.
    ioctl: No such device
    Use of uninitialized value in substitution iterator at ./check_raid.pl line 430.
    DEBUG EXEC: /sbin/mpt-status -i  at ./check_raid.pl line 449.
    ioctl: No such device
    DEBUG EXEC: /sbin/tw_cli info at ./check_raid.pl line 449.
    UNKNOWN: megacli:[Volumes(1): DISK0.0:Optimal; Devices(3): 05=Hotspare 06,07=Online]; mpt:[Plugin error]; tw_cli:[No Adapters were found on this machine]

Running `/sbin/mpt-status -p` gives:

    Checking for SCSI ID:0
    ioctl: No such device

Let me know if you need more details.

**Comment by glensc** [comment]
share the output of each of those commands please. CONTRIBUTING.md has the info how you could do that.

**Comment by glensc** [comment]
also, need to distinguish what `mpt-status` prints to `stderr`, what to `stdout`, and what is it's exit code:

```
# mpt-status -p 1>stdout 2>stderr; echo $?
```

show me above result outcome, and the contents of `stdout` and `stderr` files

**Comment by ebourg** [comment]
The other commands are unrelated to this warning, I get the same result with `check_raid.pl -p mpt` and it invokes only mpt-status.

For some reason mpt-status now returns another error message on my system, but check_raid still complains about the same uninitialized value.

Here are the outputs:

`check_raid.pl -d -p mpt` : 

    DEBUG EXEC: /sbin/mpt-status -p at ./check_raid.pl line 449.
    open /dev/mptctl: No such file or directory
      Try: mknod /dev/mptctl c 10 220
    Make sure mptctl is loaded into the kernel
    Use of uninitialized value in substitution iterator at ./check_raid.pl line 430.
    DEBUG EXEC: /sbin/mpt-status -i  at ./check_raid.pl line 449.
    open /dev/mptctl: No such file or directory
      Try: mknod /dev/mptctl c 10 220
    Make sure mptctl is loaded into the kernel
    UNKNOWN: mpt:[Plugin error]

`mpt-status -p` :

    open /dev/mptctl: No such file or directory
      Try: mknod /dev/mptctl c 10 220
    Make sure mptctl is loaded into the kernel

and `mpt-status -p 1>stdout 2>stderr; echo $?` returns 1.

**Comment by glensc** [comment]
please post those two files (`stdout` and `stderr`) as i asked. need to distinguish of those. i can't differenciate from your post what stream the output goes.

**Comment by glensc** [comment]
also, check your `lsmod` output, especially `mpt*` modules

**Comment by ebourg** [comment]
stdout is empty and stderr contains:

    open /dev/mptctl: No such file or directory
      Try: mknod /dev/mptctl c 10 220
    Make sure mptctl is loaded into the kernel

lsmod doesn't show any mpt* module

**Comment by glensc** [comment]
do you even need mpt-status there? uninstall the program so it won't initialize mpt plugin

**Comment by ebourg** [comment]
No I don't need it, but I can't uninstall it either. It comes bundled with my distribution (Proxmox).

**Comment by glensc** [comment]
i see. you can workaround by specifying plugin(s) to use with -p

**Comment by ebourg** [comment]
Yes that's what I did. But still, if someone needs mpt-status and has an error, he will get an 'uninitialized value' message instead of a proper plugin failure with a meaningful message.

**Comment by glensc** [comment]
could you try this fix? it disables the plugin if mpt-status -p does not give any controller id out

**Comment by ebourg** [comment]
I no longer get the uninitialized warning with the commit 8e11af3, thank you Elan.
